src/DataReader/SelectDataReader.php

```diff
@@ -0,0 +1,151 @@
+use Yiisoft\Yii\Cycle\DataReader\Cache\CachedCount;
+use Yiisoft\Yii\Cycle\DataReader\Cache\CachedCollection;
+
+final class SelectDataReader implements
+    DataReaderInterface,
+    OffsetableDataInterface,
+    CountableDataInterface,
+    SortableDataInterface,
+    IteratorAggregate
+{
+    /** @var Select|SelectQuery */
+    private $query;
+    private ?int $limit = null;
+    private ?int $offset = null;
+    private ?Sort $sorting = null;
+    private CachedCount $countCache;
+    private CachedCollection $itemsCache;
+
+    /**
+     * @param Select|SelectQuery $query
+     */
+    public function __construct($query)
+    {
+        if (!$query instanceof Countable) {
+            throw new InvalidArgumentException(sprintf('Query should implement %s interface', Countable::class));
+        }
+        if (!$query instanceof PaginableInterface) {
+            throw new InvalidArgumentException(
+                sprintf('Query should implement %s interface', PaginableInterface::class)
+            );
+        }
+        $this->query = clone $query;
+        $this->countCache = new CachedCount($this->query);
+        $this->itemsCache = new CachedCollection();
+    }
+
+    public function getSort(): ?Sort
+    {
+        return $this->sorting;
+    }
+
+    public function withLimit(int $limit): self
+    {
+        $clone = clone $this;
+        $clone->setLimit($limit);
+        return $clone;
+    }
+
+    public function withOffset(int $offset): self
+    {
+        $clone = clone $this;
+        $clone->setOffset($offset);
+        return $clone;
+    }
+
+    public function withSort(?Sort $sorting): self
+    {
+        $clone = clone $this;
+        $clone->setSort($sorting);
+        return $clone;
+    }
+
+    public function count(): int
+    {
+        return $this->countCache->getCount();
+    }
+
+    public function read(): iterable
+    {
+        if ($this->itemsCache->getCollection() !== null) {
+            return $this->itemsCache->getCollection();
+        }
+        $query = $this->buildQuery();
+        $this->itemsCache->setCollection($query->fetchAll());
+        return $this->itemsCache->getCollection();
+    }
+
+    private function setSort(?Sort $sorting): void
+    {
+        if ($this->sorting !== $sorting) {
+            $this->sorting = $sorting;
+            $this->itemsCache = new CachedCollection();
+        }
+    }
+
+    private function setLimit(?int $limit): void
+    {
+        if ($this->limit !== $limit) {
+            $this->limit = $limit;
+            $this->itemsCache = new CachedCollection();
+        }
+    }
+
+    private function setOffset(?int $offset): void
+    {
+        if ($this->offset !== $offset) {
+            $this->offset = $offset;
+            $this->itemsCache = new CachedCollection();
+        }
+    }
+
+    /**
+     * Get Iterator without caching
+     * @return Traversable
+     * @throws \Exception
+     */
+    public function getIterator(): Traversable
```

Comment:
Should it be `iterable` instead of `Traversable`?

Reply:
![image](https://user-images.githubusercontent.com/4152481/73866895-b1776280-4856-11ea-9c0e-e53d6e61fb96.png)